### PR TITLE
Add episode 13

### PR DESCRIPTION
**Pull request by bitsondatadev**

### CODE REVIEW

File: _episodes/13.md [hunk 1]
```diff
@@ -0,0 +1,449 @@
...
+</p>
+
+## Pinot links
+
+ * [Apache Pinot Slack](https://communityinviter.com/apps/apache-pinot/apache-pinot)
+ * [Pinot Meetup](https://www.meetup.com/apache-pinot/events/275991991/)
+
+## Guests
+  
+ * Xiang Fu, [Apache Pinot](https://pinot.apache.org/) PMC and Commiter and
```
Comment (mosabua) [inline _episodes/13.md:47]: maybe reword to 

PMC and committer at Apache Pinot, and founder of stealth mode startup

Also .. maybe we have to spell out PMC (Project Management Chair)

File: _episodes/13.md [hunk 2]
```diff
@@ -0,0 +1,449 @@
...
+## Pinot links
+
+ * [Apache Pinot Slack](https://communityinviter.com/apps/apache-pinot/apache-pinot)
+ * [Pinot Meetup](https://www.meetup.com/apache-pinot/events/275991991/)
+
+## Guests
+  
+ * Xiang Fu, [Apache Pinot](https://pinot.apache.org/) PMC and Commiter and 
+ founder of stealth mode startup ([@xiangfu0](https://twitter.com/xiangfu0))
+ * Elon Azouley, Software Engineer at stealth mode startup ([@ElonAzoulay](https://twitter.com/ElonAzoulay))
```
Comment (mosabua) [inline _episodes/13.md:49]: , software engineer at .. 

Since we use lowercase in the above one for founder and else..

File: _episodes/13.md [hunk 3]
```diff
@@ -0,0 +1,449 @@
...
+
+## Release 353
+
+Release Notes discussed: <https://trino.io/docs/current/release/release-353.html>
```
Comment (mosabua) [inline _episodes/13.md:53]: Release notes

File: _episodes/13.md [hunk 4]
```diff
@@ -0,0 +1,449 @@
...
+ - Support for correlated subqueries involving UNNEST
+ - CREATE/DROP TABLE in BigQuery connector
+ - Reading and writing column stats in Glue Metastore
+ - Support for Apache Phoenix 5.1
+
+Manfred’s notes:
+ - New geometry functions
+ - A whole bunch of correctness and performance improvements
+ - Env var (and hence secrets) support for RPM-based installs
+ - BigQuery - CREATE and DROP TABLE
```
Comment (mosabua) [inline _episodes/13.md:65]: remove this line (duplicates above)

File: _episodes/13.md [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+ - Hive - Glue stats usage
```
Comment (mosabua) [inline _episodes/13.md:66]: Remove this line (duplicate)

File: _episodes/13.md [hunk 6]
```diff
@@ -0,0 +1,449 @@
...
+ - Hive - performance for bucketed table inserts
+ - Kafka - schema registry improvements
+ - Experimental join pushdown in a bunch of JDBC connectors
+ - Also a bunch of fixes on JDBC connectors
+ - Quite a list of changes on the SPI - ensure to check if you have a plugin
+
+## Concept of the week: Data cubes and Apache Pinot
+
```
Comment (mosabua) [inline _episodes/13.md:74]: Add some sort of intro sentence..

File: _episodes/13.md [hunk 7]
```diff
@@ -0,0 +1,449 @@
...
+### Data cubes and MOLAP (Multi-dimensional online analytics processing)
+
+During data analytics, there are many access patterns that tend to repeat
```
Comment (mosabua) [inline _episodes/13.md:77]: In data analytics

File: _episodes/13.md [hunk 8]
```diff
@@ -0,0 +1,449 @@
...
+themselves over and over again. It is very common to need to split and merge
+data based on the dates and times that they occur. Or perhaps you ask a lot of
```
Comment (mosabua) [inline _episodes/13.md:79]: based on date and time values.

File: _episodes/13.md [hunk 9]
```diff
@@ -0,0 +1,449 @@
...
+questions based on a specific customer, or even a specific product. The
+questions themselves are typically have answers involving aggregate data like
```
Comment (mosabua) [inline _episodes/13.md:81]: Answering these questions typically involves aggregation of data like

File: _episodes/13.md [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+sums, averages, counts, etc... Wouldn't it make sense to cache some of these
+results?
```
Comment (mosabua) [inline _episodes/13.md:83]: intermediary results

File: _episodes/13.md [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+or range of values is to show them as a cube, that is sliced up into smaller
+dimensions. This actually derives from the traditional form of OLAP, 
+multi-dimensional OLAP (MOLAP). 
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/13/data_cube.png"/><br/>
+</p>
+
+This cube represents a caching of data aggregations that are grouped by commonly
+used dimensions. For example, the above cube would be the pre-aggregation of the
```
Comment (mosabua) [inline _episodes/13.md:95]: the displayed cube represents the pre-aggregration of the following query:

File: _episodes/13.md [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+following query:
+
+```
+SELECT part, store, customer, COUNT(*)
+FROM cube_table
+GROUP BY part, store, customer
+```
+
+If we want to get the data for a particular customer, we can take a "slice" of
+that cube by specifying a particular customer. The following query would give
```
Comment (mosabua) [inline _episodes/13.md:105]: query returns the green

File: _episodes/13.md [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+
+```
+SELECT part, store, COUNT(*)
+FROM cube_table
+WHERE customer = "Bob"
+GROUP BY part, store
+```
+
+Now what if we want to flatten one of the dimensions? While this can be managed
+with a GROUP BY as before, but depending on the system may ignore any cached
```
Comment (mosabua) [inline _episodes/13.md:116]: I think we should use backticks around SQL command like GROUP BY ..

```
`GROUP BY` as before, 
```

File: _episodes/13.md [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+data and scan over all the rows. For this, SQL reserved a special set of
+keywords around cubes. We won't dive into that in depth now but for our current
```
Comment (mosabua) [inline _episodes/13.md:118]: now, but

File: _episodes/13.md [hunk 15]
```diff
@@ -0,0 +1,449 @@
...
+goal of flattening a dimension, we can use ROLLUP. Using the keyword ROLLUP
+indicates to the underlying system that you intend to aggregate over the 
+pre-materialized data rather than scan over all rows to compute again. This will
```
Comment (mosabua) [inline _episodes/13.md:121]: This gives you the total

File: _episodes/13.md [hunk 16]
```diff
@@ -0,0 +1,449 @@
...
+
+```
+SELECT part, store, COUNT(*)
+FROM cube_table
+GROUP BY ROLLUP (part, store)
+```
+
+Now, although we used simple counts, you can precompute a lot of other aggregate
+data like sums, min, max, percentile, etc... These can service various queries
+that are commonly queried and doesn't require a new computation every time. That
```
Comment (mosabua) [inline _episodes/13.md:132]: and don't require

File: _episodes/13.md [hunk 17]
```diff
@@ -0,0 +1,449 @@
...
+is the goal of MOLAP and data cubes.
+
+### Apache Pinot
+
+Now let's move on to Apache Pinot is a Realtime distributed OLAP datastore,
```
Comment (mosabua) [inline _episodes/13.md:137]: Pinot. It is a realtime distributed OLAP ..

File: _episodes/13.md [hunk 18]
```diff
@@ -0,0 +1,449 @@
...
+
+Now let's move on to Apache Pinot is a Realtime distributed OLAP datastore, 
+designed to answer OLAP queries with low latency. Although there may be a lot of
+words there that overlap with the Trino description, the key differentiators are
+realtime and low latency. Trino performs batch processing and is not a realtime
+system where Pinot is great for ingesting data in batch or stream. The other key
+word, low latency could technically apply to both Pinot and Trino but in the
+context of realtime subsecond latency, Trino is slow compared to Pinot. This
+is due to the specialized indexes that Pinot uses to store the data that we
+will cover shortly. Xiang has a really great summary slide that easily shows
```
Comment (mosabua) [inline _episodes/13.md:145]: that we cover shortly. Also importantly Trino does not store any data itself. It is purely a query engine.

File: _episodes/13.md [hunk 19]
```diff
@@ -0,0 +1,449 @@
...
+the strengths of each system and why they work so well together.
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/13/latency_flexibility.png"/><br/>
+</p>
+
+While Trino is not as fast as Pinot, it is able to handle a broader set of
+use cases like performing broad joins over open data formats in data lakes. 
+This is what motivated work on the Trino Pinot Connector. You can have the speed
```
Comment (mosabua) [inline _episodes/13.md:154]: connector

File: _episodes/13.md [hunk 20]
```diff
@@ -0,0 +1,449 @@
...
+of Pinot, while having the flexibilty of Trino.
+
+Now that you understand the common use case for Pinot, it's important to know 
+the main goals of Pinot. 
+ 1. One primary goal is the keep response times of aggregation queries
```
Comment (mosabua) [inline _episodes/13.md:159]: empty line and no indent for 1.

File: _episodes/13.md [hunk 21]
```diff
@@ -0,0 +1,449 @@
...
+  predictable, regardless of how many requests Pinot handles. As it scales
+  you won't see a degradation of performance. This is achieved by Pinot's
+  custom indices and storage formats. 
+    <p align="center">
+       <img align="center" width="75%" height="100%" src="/assets/episode/13/data_value.jpeg"/><br/>
+    </p>
+
+ 2. Another goal of Pinot is to revive the value of data from a historical
```
Comment (mosabua) [inline _episodes/13.md:167]: revive ?

File: _episodes/13.md [hunk 22]
```diff
@@ -0,0 +1,449 @@
...
+  context. Data will reach a particular point in its lifecycle where it
```
Comment (mosabua) [inline _episodes/13.md:168]: Data reaches

File: _episodes/13.md [hunk 23]
```diff
@@ -0,0 +1,449 @@
...
+  becomes less valuable as it ages. While all data is able to add some value no
+  matter what the age, there's a tradeoff of scanning multiple rows to glean
+  information from antiquated data. Pinot aims to remove this tradeoff as most 
+  questions around historical data are queried in aggregate and this can be
+  summarized and queried at a low cost.
+ 
+ 3. The final goal is to manage dimension explosion. One of the difficulties
+  with managing a system that caches all this historic data is handling
+  dimension explosion that occurs when you cache every possibly combination of
```
Comment (mosabua) [inline _episodes/13.md:177]: possible

File: _episodes/13.md [hunk 24]
```diff
@@ -0,0 +1,449 @@
...
+  data. Above we showed a 3 dimensional cube, but Pinot can handle a much
```
Comment (mosabua) [inline _episodes/13.md:178]: three-dimensional

File: _episodes/13.md [hunk 25]
```diff
@@ -0,0 +1,449 @@
...
+  larger number of dimensions. However, just because you can, doesn't mean you
+  should. Pinot has a lot of smarts around using the data, and some good
+  defaults to determine the maximum number of buckets per dimension. This helps
+  balance an exploding cache yet maintains fast results. 
+
+### Pinot Architecture
```
Comment (mosabua) [inline _episodes/13.md:184]: architecture

File: _episodes/13.md [hunk 26]
```diff
@@ -0,0 +1,449 @@
...
+
+We just covered Pinot theory and goals, let's take a quick look at their
```
Comment (mosabua) [inline _episodes/13.md:186]: at the architecture:

File: _episodes/13.md [hunk 27]
```diff
@@ -0,0 +1,449 @@
...
+[broker](https://docs.pinot.apache.org/basics/components/broker), 
+[server](https://docs.pinot.apache.org/basics/components/server), and
+optionally a [minion](https://docs.pinot.apache.org/basics/components/minion)
+to purge data.
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/13/pinot_architecture.svg"/><br/>
+</p>
+
+## PR of the week: PR 2028 Add pinot connector
```
Comment (mosabua) [inline _episodes/13.md:200]: Add Pinot connector

File: _episodes/13.md [hunk 28]
```diff
@@ -0,0 +1,449 @@
...
+
+In this week's pull request <https://github.com/trinodb/trino/pull/2028> that
```
Comment (mosabua) [inline _episodes/13.md:202]: Our guest on the show today, Elon Azouley, is the author of this PR, so we can ask him all about it now.

File: _episodes/13.md [hunk 29]
```diff
@@ -0,0 +1,449 @@
...
+was graciously added by our guest, Elon.
+
+<p align="center">
+<img align="center" width="75%" height="100%" src="/assets/episode/13/trino_pinot_connector.png"/><br/>
+</p>
+
+ - [Basic configuration (pinot controller url, pinot segment limit)](https://trino.io/docs/current/connector/pinot.html#configuration)
```
Comment (mosabua) [inline _episodes/13.md:209]: Pinot controller URL, Pinot segment limit)

File: _episodes/13.md [hunk 30]
```diff
@@ -0,0 +1,449 @@
...
+ - 2 ways to connect to pinot - broker and server, and their tradeoffs
```
Comment (mosabua) [inline _episodes/13.md:210]: Pinot

File: _episodes/13.md [hunk 31]
```diff
@@ -0,0 +1,449 @@
...
+ 
+ - [Pinot insert (PR 7162)](https://github.com/trinodb/trino/pull/7162)
+ - [Pinot create table (PR 7164)](https://github.com/trinodb/trino/pull/7164)
+ - [Pinot drop table (PR 7160)](https://github.com/trinodb/trino/pull/7160)
+ - [Pinot 6 (PR 7163)](https://github.com/trinodb/trino/pull/7163)
+ - Pinot filter clause parsing (see question of the week below)
+
+## Demo: Pinot batch insertion and query using Trino Pinot connector
+
+To put this PR to the test, we need to set up a Pinot cluster using Docker
```
Comment (mosabua) [inline _episodes/13.md:235]: we set up a Pinot cluster


(dont need to do it that way ;-))

File: _episodes/13.md [hunk 32]
```diff
@@ -0,0 +1,449 @@
...
+Compose.
+
+To load the data, we're going to use a simple batch import, but you can also 
+[insert the data in a stream](https://docs.pinot.apache.org/basics/data-import/upsert)
+using [Kafka](https://kafka.apache.org/).
+
+Let's start up the Pinot cluster along with the required Zookeeper and Kafka
+broker. Clone this repository and navigate
+```
```
Comment (mosabua) [inline _episodes/13.md:244]: empty line above snippet

File: _episodes/13.md [hunk 33]
```diff
@@ -0,0 +1,449 @@
...
+git clone git@github.com:bitsondatadev/trino-getting-started.git
+
+cd pinot/trino-pinot
+
+docker-compose up -d
+```
+To do batch insert, we will stage a csv file to read the data in. Create a
```
Comment (mosabua) [inline _episodes/13.md:251]: empty line below snippet

File: _episodes/13.md [hunk 34]
```diff
@@ -0,0 +1,449 @@
...
+mkdir -p /tmp/pinot-quick-start/rawdata
+
+echo "studentID,firstName,lastName,gender,subject,score,timestampInEpoch
+200,Lucy,Smith,Female,Maths,3.8,1570863600000
+200,Lucy,Smith,Female,English,3.5,1571036400000
+201,Bob,King,Male,Maths,3.2,1571900400000
+202,Nick,Young,Male,Physics,3.6,1572418800000" > /tmp/pinot-quick-start/rawdata/transcript.csv
+```
+
+In order for Pinot to understand the csv data, we must provide it a
```
Comment (mosabua) [inline _episodes/13.md:264]: CSV

File: _episodes/13.md [hunk 35]
```diff
@@ -0,0 +1,449 @@
...
+    \"tenants\" : {
+      \"broker\":\"DefaultTenant\",
+      \"server\":\"DefaultTenant\"
+    },
+    \"tableType\":\"OFFLINE\",
+    \"metadata\": {}
+}" > /tmp/pinot-quick-start/transcript-table-offline.json
+```
+
+Once you create these 3 files and verify that docker containers are running,
```
Comment (mosabua) [inline _episodes/13.md:334]: three files

File: _episodes/13.md [hunk 36]
```diff
@@ -0,0 +1,449 @@
...
+we can now run the `Add Table` command
```
Comment (mosabua) [inline _episodes/13.md:335]: command:

File: _episodes/13.md [hunk 37]
```diff
@@ -0,0 +1,449 @@
...
+    -v /tmp/pinot-quick-start:/tmp/pinot-quick-start \
+    --name pinot-batch-table-creation \
+    apachepinot/pinot:latest AddTable \
+    -schemaFile /tmp/pinot-quick-start/transcript-schema.json \
+    -tableConfigFile /tmp/pinot-quick-start/transcript-table-offline.json \
+    -controllerHost pinot-controller \
+    -controllerPort 9000 -exec
+```
+
+Now, we created a table, we can actually see our table exists in the
```
Comment (mosabua) [inline _episodes/13.md:349]: Now that the table exists, we can see it in the Pinot web UI

File: _episodes/13.md [hunk 38]
```diff
@@ -0,0 +1,449 @@
...
+    --name pinot-batch-table-creation \
+    apachepinot/pinot:latest AddTable \
+    -schemaFile /tmp/pinot-quick-start/transcript-schema.json \
+    -tableConfigFile /tmp/pinot-quick-start/transcript-table-offline.json \
+    -controllerHost pinot-controller \
+    -controllerPort 9000 -exec
+```
+
+Now, we created a table, we can actually see our table exists in the 
+[Pinot web ui](http://localhost:9000/#/tables). Now we want to actually want to insert data
```
Comment (mosabua) [inline _episodes/13.md:350]: Let's insert some data using a batch job specification:

File: _episodes/13.md [hunk 39]
```diff
@@ -0,0 +1,449 @@
...
+  className: 'org.apache.pinot.plugin.inputformat.csv.CSVRecordReader'
+  configClassName: 'org.apache.pinot.plugin.inputformat.csv.CSVRecordReaderConfig'
+tableSpec:
+  tableName: 'transcript'
+  schemaURI: 'http://pinot-controller:9000/tables/transcript/schema'
+  tableConfigURI: 'http://pinot-controller:9000/tables/transcript'
+pinotClusterSpecs:
+  - controllerURI: 'http://pinot-controller:9000'" > /tmp/pinot-quick-start/docker-job-spec.yml
+```
+Now run this batch job by running the `LaunchDataIngestionJob` task.
```
Comment (mosabua) [inline _episodes/13.md:378]: Empty line above and below

**Review by mosabua — CHANGES_REQUESTED** [review]
A bunch of changes...

File: _episodes/13.md (hunk 21 above)
Comment (bitsondatadev) [reply _episodes/13.md:167]: One of Xiang's points is that data loses it's value as time goes on, especially ones it falls of the grid of fast searching data stores like elasticsearch/redis/mongo etc.. Using Pinot as a fast way to serve aggregate data revives that value that would otherwise be lost in historic data vs having to scan over all the old data every time to get it's value.

We don't have to use revive but that's the jist of what I'm trying to explain.

File: _episodes/13.md (hunk 27 above)
Comment (bitsondatadev) [reply _episodes/13.md:200]: This was verbatim of the title used in the PR. I guess we can doctor it up.

File: _episodes/13.md (hunk 21 above)
Comment (mosabua) [reply _episodes/13.md]: increase the value of historical data

or something like that maybe

**Review by mosabua — APPROVED** [review]
Looks good to go now.

File: _episodes/13.md [hunk 40]
```diff
@@ -0,0 +1,479 @@
...
+</p>
+
+## Pinot links
+
+ * [Apache Pinot Slack](https://communityinviter.com/apps/apache-pinot/apache-pinot)
+ * [Pinot Meetup](https://www.meetup.com/apache-pinot/events/275991991/)
+
+## Guests
+  
+ * Xiang Fu, project management chair and commiter at [Apache Pinot](https://pinot.apache.org/)
```
Comment (martint) [inline _episodes/13.md:50]: Typo: Committer

File: _episodes/13.md [hunk 41]
```diff
@@ -0,0 +1,479 @@
...
+  and co-founder of stealth mode startup ([@xiangfu0](https://twitter.com/xiangfu0))
+ * Elon Azouley, software engineer at stealth mode startup ([@ElonAzoulay](https://twitter.com/ElonAzoulay))
```
Comment (martint) [inline _episodes/13.md:52]: Typo: Azoulay

**Comment by bitsondatadev** [comment]
Fixed those, going to do one more sweep to see if there's anything standing out.

**Comment by bitsondatadev** [comment]
Okay fixed and did another typo sweep.
